<a href="https://colab.research.google.com/github/jblanco89/ML-Zoomcamp-course/blob/main/ML_Zoomcamp_homework_w1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML Zoomcamp Homework
## Week 1
### Module 1 - intro
*Author*: Javier Blanco

In [ ]:
#Import necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as ptl
%matplotlib inline

## Question 1
What's the version of Pandas that you installed?

In [ ]:
pd.__version__

'1.5.3'

## Getting the Data
we'll use the California Housing Prices dataset.

In [ ]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv

--2023-09-11 14:17:28--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1423529 (1.4M) [text/plain]
Saving to: ‘housing.csv.2’

housing.csv.2       100%[===================>]   1.36M  --.-KB/s    in 0.08s   

2023-09-11 14:17:28 (17.5 MB/s) - ‘housing.csv.2’ saved [1423529/1423529]



In [ ]:
df = pd.read_csv('/content/housing.csv', sep=',')
df.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563400   
50%        435.000000   1166.000000    409.000000       3.534800   
75%        647.000000   1725.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

In [ ]:
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

## Question 2
How many columns are in the dataset?

1.   10
2.   6560
3.   10989
4.   20640


In [ ]:
df.shape[1]

10

## Question 3
Which columns in the dataset have missing values?

1.   `total_rooms`
2.   `total_bedrooms`
3.   both of the above
4.   no empty columns in the dataset


In [ ]:
df.isna().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [ ]:
df.columns[df.isna().any()].tolist()

['total_bedrooms']

## Question 4
How many unique values does the `ocean_proximity` column have?

1.   3
2.   5
3.   7
4.   9


In [ ]:
len(df['ocean_proximity'].unique())

5

## Question 5
What's the average value of the `median_house_value` for the houses located near the bay?

1.   49433
2.   124805
3.   259212
4.   380440


In [ ]:
df['median_house_value'][df['ocean_proximity'] == 'NEAR BAY'].mean()

259212.31179039303

## Question 6

1.   Calculate the average of
total_bedrooms column in the dataset.
2.   Use the fillna method to fill the missing values in total_bedrooms with the mean value from the previous step.
3.   Now, calculate the average of total_bedrooms again.
4. Has it changed?

In [ ]:
#before fillna
mean_before_na = df['total_bedrooms'].mean()
mean_before_na

537.8705525375618

In [ ]:
#filling nan
total_bedrooms_not_na = df['total_bedrooms'].fillna(mean_before_na)
total_bedrooms_not_na

0         129.0
1        1106.0
2         190.0
3         235.0
4         280.0
          ...  
20635     374.0
20636     150.0
20637     485.0
20638     409.0
20639     616.0
Name: total_bedrooms, Length: 20640, dtype: float64

In [ ]:
mean_after_na = total_bedrooms_not_na.mean()
mean_after_na

537.8705525375617

**NO**

## Question 7

1. Select all the options located on islands.
2. Select only columns housing_median_age, total_rooms, total_bedrooms.
3. Get the underlying NumPy array. Let's call it X.
4. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

5. Compute the inverse of XTX.
6. Create an array y with values [950, 1300, 800, 1000, 1300].
7. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

What's the value of the last element of w?

*   -1.4812
*   0.001
*   5.6992
*   23.1233


In [ ]:
X = df[['housing_median_age', 'total_rooms','total_bedrooms']][df['ocean_proximity'] == 'ISLAND'].to_numpy()
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [ ]:
X.T

array([[  27.,   52.,   52.,   52.,   29.],
       [1675., 2359., 2127.,  996.,  716.],
       [ 521.,  591.,  512.,  264.,  214.]])

In [ ]:
XTX = np.matmul(X.T, X)
XTX

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [ ]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [ ]:
y = np.array([950, 1300, 800, 1000, 1300])
y

array([ 950, 1300,  800, 1000, 1300])

In [ ]:
inv_mult = np.matmul(XTX_inv, X.T)
inv_mult

array([[-0.00825608, -0.00653208, -0.00232159,  0.02565144,  0.01204934],
       [-0.00165852,  0.0011141 ,  0.00139656, -0.00103215, -0.00110698],
       [ 0.00754365, -0.00301964, -0.00455125,  0.00181685,  0.00329418]])

In [ ]:
w = inv_mult.dot(y)
w[2]

5.699229455065586